### PR TITLE
Topic Proxy: leave request handling

### PR DESCRIPTION
**Pull request by aforge**
Pass over session leave requests to the topic master and
clean up master & proxy topics (and their helper sessions) when there are no more attached sessions to them.

Note that this PR is an intermediate version. There's more work to be done to property handle subscribe/unsubscribe.

### CODE REVIEW

File: server/topic.go [hunk 1]
```diff
@@ -2956,19 +3007,28 @@ func (t *Topic) addSession(s *Session, asUid types.Uid) bool {
...
-// Removes session record if 'asUid' matches subscribed user.
-func (t *Topic) remSession(s *Session, asUid types.Uid) *perSessionData {
-	if pssd, ok := t.sessions[s]; ok && (pssd.uid == asUid || asUid.IsZero()) {
-		// Check for deferred presence notification and cancel it if found.
-		if pssd.ref != nil {
-			t.defrNotif.Remove(pssd.ref)
+// Removes session record if 'doRemove' is true and either:
+// * 'asUid' matches subscribed user.
+// * or 's' is a proxy session for a proxy-master topic connection.
+func (t *Topic) maybeRemoveSession(s *Session, asUid types.Uid, doRemove bool) *perSessionData {
```
Comment (or-else) [inline server/topic.go:3013]: The logic here is for the use case when the session authenticated with a root user A is subscribed to topic on behalf of user B. It also implements support for deferred online notifications. The `pssd` (per session data) contains user id overrides + reference to deferred notification. 

The session is removed from topic when the actually subscribed user  (A==B) leaves `pssd.uid == asUid` or session is disconnected `asUid.IsZero()` regardless of user (similar to `leave.terminateProxyConnection == true`).

If session is removed from topic then non-nil `pssd` is returned. Otherwise nil is returned.

How it applies to a session from a proxy topic:

* Proxy session at master topic should have zeros and nils in `pssd`.
* `pssd` should not be checked at the master topic (when `s.isProxy` is `true`). `pssd` is for supporting onBehalfOf. This session is always `onBehalfOf` multiple users. It should be checked at proxy topic (`t.isProxy` is `true`).
* The `t.defrNotif.Remove(pssd.ref)` should not be called at the master topic. It should be called  at the origin, at the proxy topic.
* Basically, the proxy session should be unconditionally removed when the origin terminates it. Otherwise it should not be affected.

File: server/topic.go [hunk 2]
```diff
@@ -338,22 +381,28 @@ func (t *Topic) runLocal(hub *Hub) {
...
 				// User wants to leave and unsubscribe.
 				// asUid must not be Zero.
-				if err := t.replyLeaveUnsub(hub, leave.sess, asUid, leave.id, nil/*leave.sessOverrides*/); err != nil {
+				if err := t.replyLeaveUnsub(hub, leave.sess, asUid, leave.id, leave.sessOverrides); err != nil {
 					log.Println("failed to unsub", err, leave.sess.sid)
 					continue
 				}
 
-			} else if pssd := t.remSession(leave.sess, asUid); pssd != nil {
+			} else if pssd := t.maybeRemoveSession(leave.sess, asUid, /*doRemove=*/!leave.sess.isProxy || leave.terminateProxyConnection); pssd != nil {
```
Comment (or-else) [inline server/topic.go:396]: See comment at declaration of `maybeRemoveSession()`.

File: server/topic.go [hunk 3]
```diff
@@ -338,22 +381,28 @@ func (t *Topic) runLocal(hub *Hub) {
...
 				// Just leaving the topic without unsubscribing if user is subscribed.
 
-				pud := t.perUser[pssd.uid]
+				var uid types.Uid
+				if t.isProxy {
```
Comment (or-else) [inline server/topic.go:400]: This section is probably incorrect. 

* Behavior at the [proxy topic + normal session] should match that of the [normal topic + normal session]:
`pud := t.perUser[pssd.uid]` like before.
* Behavior of [normal topic + proxy session] should be different: it should not care abut updating any of `t.perUser[uid] = pud` because proxy session multiplexes many uses. 

See comment at declaration of `maybeRemoveSession()`.

File: server/cluster.go [hunk 4]
```diff
@@ -511,6 +526,7 @@ func (c *Cluster) TopicMaster(msg *ClusterReq, rejected *bool) error {
 		// If the session is not found, create it.
 		var count int
 		sess, count = globals.sessionStore.NewSession(node, sid)
+    sess.isProxy = true
```
[the remark this answers is not in the document]
Comment (aforge) [reply server/cluster.go:529]: Done.

**Comment by or-else** [comment]
I have a suggestion for improving the review process. If you don't mind, please merge `devel` into `topic-proxy` directly instead of merging it into `tp-dev`. Otherwise my changes to `devel` are mixed with your changes to `tp-dev`.

File: server/topic.go (hunk 3 above)
Comment (or-else) [reply server/topic.go]: OK
